pyrealm/demography/canopy_functions.py

```diff
@@ -0,0 +1,411 @@
+    following shapes:
+
+    1. A scalar array: :math:`q(z)` is found for all stems at the same height and the
+       return value is a 1D array of length :math:`I`.
+    2. A row vector of length :math:`I`: :math:`q(z)` is found for all stems at
+       stem-specific heights and the return value is again a 1D array of length
+       :math:`I`.
+    3. A column vector of length :math:`J`, that is a 2 dimensional array of shape
+       (:math:`J`, 1). This allows :math:`q(z)` to be calculated efficiently for a set
+       of heights for all stems and return a 2D array of shape (:math:`J`, :math:`I`).
+
+    Args:
+        z: Height at which to calculate relative radius
+        stem_height: Total height of individual stem
+        m: Canopy shape parameter of PFT
+        n: Canopy shape parameter of PFT
+        validate: Boolean flag to suppress argument validation.
+    """
+
+    if validate:
+        _validate_z_args(z, stem_height, m, n)
+
+    z_over_height = z / stem_height
+
+    return m * n * z_over_height ** (n - 1) * (1 - z_over_height**n) ** (m - 1)
+
+
+def calculate_stem_projected_crown_area_at_z(
+    z: NDArray[np.float32],
+    q_z: NDArray[np.float32],
+    stem_height: NDArray[np.float32],
+    crown_area: NDArray[np.float32],
+    q_m: NDArray[np.float32],
+    z_max: NDArray[np.float32],
+    validate: bool = True,
+) -> NDArray[np.float32]:
+    """Calculate stem projected crown area above a given height.
+
+    This function calculates the projected crown area of a set of stems with given
+    properties at a set of vertical heights. The stem properties are given in the
+    arguments ``stem_height``,``crown_area``,``q_m`` and ``z_max``, which must be
+    one-dimensional arrays ('row vectors') of equal length. The array of vertical
+    heights ``z`` accepts a range of input shapes (see
+    :meth:`~pyrealm.demography.canopy_functions.calculate_relative_canopy_radius_at_z`
+    ) and this function then also requires the expected relative stem radius (``q_z``)
+    calculated from those heights.
+
+    Args:
+        z: Vertical height at which to estimate crown area
+        q_z: Relative crown radius at those heights
+        crown_area: Crown area of each stem
+        stem_height: Stem height of each stem
+        q_m: Canopy shape parameter ``q_m``` for each stem
+        z_max: Height of maximum crown radous for each stem
+        validate: Boolean flag to suppress argument validation.
+    """
+
+    if validate:
+        _validate_z_args(z, stem_height, crown_area, q_m, z_max)
+        _validate_q_z(z, q_z, crown_area)
+
+    # Calculate A_p
+    # Calculate Ap given z > zm
+    A_p = crown_area * (q_z / q_m) ** 2
+    # Set Ap = Ac where z <= zm
+    A_p = np.where(z <= z_max, crown_area, A_p)
+    # Set Ap = 0 where z > H
+    A_p = np.where(z > stem_height, 0, A_p)
+
+    return A_p
+
+
+def solve_community_projected_canopy_area(
+    z: float,
```

Comment:
Sometimes a really long argument list can be a bit cumbersome so worth considering dataclass or namedtuple for such cases. However, that comes with different overheads so a bit of a style choice.

Reply:
This specific module has a bunch of extremely similar but not identical signatures. We could bundle everything up in one dataclass (and indeed @AmyOctoCat had these using `Community` for exactly that reason), but I'd like to retain the flexibility of using these directly.

One reason for that is that I want to build some user facing helper classes to generate canopy data outside of a Community object, so it's probably easier to keep them like this, but I will be looking at that next.